core/src/main/scala/org/apache/spark/api/python/VirtualEnvFactory.scala

```diff
@@ -0,0 +1,151 @@
+  private var isLauncher: Boolean = false
+
+  // used by launcher when user want to use virtualenv in pyspark shell. Launcher need this class
+  // to create virtualenv for driver.
+  def this(pythonExec: String, properties: JMap[String, String], isDriver: java.lang.Boolean) {
+    this(pythonExec, new SparkConf(), isDriver)
+    properties.asScala.foreach(entry => this.conf.set(entry._1, entry._2))
+    virtualEnvType = conf.get("spark.pyspark.virtualenv.type", "native")
+    virtualEnvPath = conf.get("spark.pyspark.virtualenv.bin.path")
+    this.isLauncher = true
+  }
+
+  /*
+   * Create virtualenv using native virtualenv or conda
+   *
+   * Native Virtualenv:
+   *   -  Execute command: virtualenv -p <pythonExec> --no-site-packages <virtualenvName>
+   *   -  Execute command: python -m pip --cache-dir <cache-dir> install -r <requirement_file>
+   *
+   * Conda
+   *   -  Execute command: conda create --prefix <prefix> --file <requirement_file> -y
+   *
+   */
+  def setupVirtualEnv(): String = {
+    logInfo("Start to setup virtualenv...")
+    logDebug("user.dir=" + System.getProperty("user.dir"))
+    logDebug("user.home=" + System.getProperty("user.home"))
+
+    require(virtualEnvType == "native" || virtualEnvType == "conda",
+      s"VirtualEnvType: ${virtualEnvType} is not supported." )
+    require(new File(virtualEnvPath).exists(),
+      s"VirtualEnvPath: ${virtualEnvPath} is not defined or doesn't exist.")
+    // Use a temp directory for virtualenv in the following cases:
+    // 1. driver of pyspark shell
+    // 2. driver of yarn-client mode
+    // otherwise create the virtualenv folder under the executor working directory.
+    if (isLauncher ||
+      (isDriver && conf.get("spark.submit.deployMode") == "client")) {
+      val virtualenv_basedir = Files.createTempDir()
+      virtualenv_basedir.deleteOnExit()
+      virtualEnvName = virtualenv_basedir.getAbsolutePath
+    } else if (isDriver && conf.get("spark.submit.deployMode") == "cluster") {
+      virtualEnvName = "virtualenv_driver"
+    } else {
+      // use the working directory of Executor
+      virtualEnvName = "virtualenv_" + conf.getAppId + "_" + VIRTUALENV_ID.getAndIncrement()
+    }
+
+    // Use the absolute path of requirement file in the following cases
+    // 1. driver of pyspark shell
+    // 2. driver of yarn-client mode
+    // otherwise just use filename as it would be downloaded to the working directory of Executor
+    val pyspark_requirements =
+      if (isLauncher ||
+        (isDriver && conf.get("spark.submit.deployMode") == "client")) {
+        conf.getOption("spark.pyspark.virtualenv.requirements")
+      } else {
+        conf.getOption("spark.pyspark.virtualenv.requirements").map(_.split("/").last)
+      }
+
+    val createEnvCommand =
+      if (virtualEnvType == "native") {
+        Arrays.asList(virtualEnvPath,
+          "-p", pythonExec,
+          "--no-site-packages", virtualEnvName)
```

Comment:
According to the reference guide of `virtualenv`, `--no-site-packages` is deprecated. Do we still need this? https://virtualenv.pypa.io/en/stable/reference/#cmdoption-no-site-packages